Godeps/_workspace/src/github.com/jvatic/asset-matrix-go/cache.go

```diff
@@ -0,0 +1,143 @@
+package assetmatrix
+
+import (
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+type Cache struct {
+	Dir            string
+	cacheHits      []string
+	cacheHitsMutex sync.Mutex
+}
+
+func (c *Cache) FindCachedAsset(a Asset) (Asset, error) {
+	checkSum := a.CheckSum()
+	if checkSum == "" {
+		return a, nil
+	}
+	p := filepath.Join(c.Dir, checkSum)
+	if _, err := os.Lstat(p); err != nil {
+		return nil, err
+	}
+	c.cacheHitsMutex.Lock()
+	c.cacheHits = append(c.cacheHits, p)
+	c.cacheHitsMutex.Unlock()
+	return &CachedAsset{input: a, p: p}, nil
+}
+
+func (c *Cache) CacheAsset(ar io.Reader, checkSum string) io.Reader {
+	if checkSum == "" {
+		return ar
+	}
+	p := filepath.Join(c.Dir, checkSum)
+	if err := os.MkdirAll(c.Dir, 0755); err != nil {
+		// abort
+		log.Println(err)
+		return ar
+	}
+	f, err := os.OpenFile(p, os.O_CREATE|os.O_WRONLY, 0755)
+	if err != nil {
+		// abort
+		log.Println(err)
+		return ar
+	}
+	c.cacheHitsMutex.Lock()
+	c.cacheHits = append(c.cacheHits, p)
+	c.cacheHitsMutex.Unlock()
+	cr, w := io.Pipe()
+	r := io.TeeReader(ar, w)
+	go func() {
+		defer f.Close()
+		defer w.Close()
+		if _, err := io.Copy(f, cr); err != nil {
+			log.Fatal(err)
+		}
+	}()
+	return r
+}
+
+func (c *Cache) CleanupCacheDir() error {
+	c.cacheHitsMutex.Lock()
+	defer c.cacheHitsMutex.Unlock()
+	return filepath.Walk(c.Dir, func(path string, info os.FileInfo, err error) error {
+		if path == c.Dir {
+			return nil
+		}
+		found := false
+		for _, p := range c.cacheHits {
+			if p == path {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return os.Remove(path)
+		}
+		return nil
+	})
+}
+
+type CachedAsset struct {
+	p     string
+	input Asset
+}
+
+func (a *CachedAsset) Open() (*os.File, error) {
+	return os.Open(a.p)
+}
+
+func (a *CachedAsset) Initialize() error {
+	return nil
+}
+
+func (a *CachedAsset) CheckSum() string {
+	return a.input.CheckSum()
+}
+
+func (a *CachedAsset) Path() string {
+	return a.input.Path()
+}
+
+func (a *CachedAsset) RelPath() (string, error) {
+	return a.input.RelPath()
+}
+
+func (a *CachedAsset) SetIndexKey(key string) {
+	a.input.SetIndexKey(key)
+}
+
+func (a *CachedAsset) IndexKey() string {
+	return a.input.IndexKey()
+}
+
+func (a *CachedAsset) ImportPaths() []string {
+	return a.input.ImportPaths()
+}
+
+func (a *CachedAsset) Compile() (io.Reader, error) {
+	file, err := os.Open(a.p)
+	if err != nil {
+		return nil, err
+	}
+	r, w := io.Pipe()
+	go func() {
+		defer file.Close()
+		defer w.Close()
+		if _, err := io.Copy(w, file); err != nil {
+			log.Fatal(err)
+		}
+	}()
+	return r, nil
```

Comment:
Yes, the code upstream doesn't handle closing the reader (should probably refactor it to handle that at some point).